### PR TITLE
Expose data defined control over raster layer opacity

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows the opacity of a raster to be varied e.g. on different pages of an atlas, depending on the visibility of another layer,
by temporal variables, etc...

Also add a "redraw layer only" temporal mode for raster layers. Just like the equivalent setting for vector layers, this option
causes a raster layer to be redrawn on each new animation frame. It's useful when the layer uses time-based expression values (e.g. data defined renderer opacity, to fade/in out a raster layer)

Together this allows for this kind of effect:

https://user-images.githubusercontent.com/1829991/123203489-cf03a580-d4f9-11eb-97b6-f684a139c5c3.mp4

### CODE REVIEW

File: src/gui/raster/qgsrasterlayerproperties.h [hunk 1]
```diff
@@ -85,12 +87,40 @@ class GUI_EXPORT QgsRasterLayerProperties : public QgsOptionsDialogBase, private
...
     void optionsStackedWidget_CurrentChanged( int index ) override SIP_SKIP ;
 
+  private:
+
+    // TODO -- consider moving these to a common raster widget base class
+
+    /**
+     * Registers a property override button, setting up its initial value, connections and description.
+     * \param button button to register
+     * \param key corresponding data defined property key
```
Comment (nirvn) [inline src/gui/raster/qgsrasterlayerproperties.h:103]: \since QGIS 3.22

File: src/gui/raster/qgsrasterlayerproperties.cpp [hunk 2]
```diff
@@ -1582,6 +1602,43 @@ void QgsRasterLayerProperties::optionsStackedWidget_CurrentChanged( int index )
   }
 }
 
+void QgsRasterLayerProperties::initializeDataDefinedButton( QgsPropertyOverrideButton *button, QgsRasterPipe::Property key )
```
Comment (nirvn) [inline src/gui/raster/qgsrasterlayerproperties.cpp:1605]: Do you also need to add code to handle the styling panel?

Comment (nyalldawson) [reply src/gui/raster/qgsrasterlayerproperties.cpp:1605]: Can you clarify?

**Review by nirvn — APPROVED** [review]

File: src/gui/raster/qgsrasterlayerproperties.h (hunk 1 above)
Comment (nyalldawson) [reply src/gui/raster/qgsrasterlayerproperties.h:103]: Ah, it's private methods, but I've added it anyway

File: src/gui/raster/qgsrasterlayerproperties.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/gui/raster/qgsrasterlayerproperties.cpp:1605]: (actually it's already handled in this pr)

**Pull request merged**